### PR TITLE
os_mon: allow check intervals smaller than 1 minute

### LINKED ISSUE #6283: "disksup" should allow check intervals smaller than 1 minute
**Is your feature request related to a problem? Please describe.**

Team RabbitMQ would love to use the built-in disk monitor that ships with Erlang/OTP - `disksup`. Unfortunately, the smallest `disk_space_check_interval` allowed is 1 minute. RabbitMQ users can be very good a filling disks so we have written our own disk monitor that allows more frequent checks. I'm certain other data services would benefit from shorter check intervals as well.

**Describe the solution you'd like**

We would like to be able to specify `disk_space_check_interval` values in milliseconds. Currently, the default minimum value allowed in RabbitMQ is 100ms.

**Describe alternatives you've considered**

RabbitMQ has its own disk monitor to work around this, but we'd love to use `disksup`!

**Additional context**

Related issue - https://github.com/erlang/otp/issues/6156

Maintaining our own disk monitor has been a source of headaches -
* https://github.com/rabbitmq/rabbitmq-server/pull/5241
* https://github.com/rabbitmq/rabbitmq-server/pull/5726

### PR DESCRIPTION
**Pull request by jdamanalo**
- Expanded type of `disk_space_check_interval` to `number()`

```
1> disksup:set_check_interval(0.5).
```

Fix #6283

### CODE REVIEW

File: lib/os_mon/src/disksup.erl [hunk 1]
```diff
@@ -77,8 +77,8 @@ dummy_reply(get_almost_full_threshold) ->
 dummy_reply({set_almost_full_threshold, _}) ->
     ok.
 
-param_type(disk_space_check_interval, Val) when is_integer(Val),
-						Val>=1 -> true;
+param_type(disk_space_check_interval, Val) when is_number(Val),
+						0<Val -> true;
```
Comment (gomoripeti) [inline lib/os_mon/src/disksup.erl:81]: not sure if this is a concern, but this seems to allow an effective check interval of zero if `Val` is smaller than `1 / 60_000`

Comment (jdamanalo) [reply lib/os_mon/src/disksup.erl]: thanks

**Comment by lukebakken** [comment]
@jdamanalo thanks for getting this, it was my next task after #6384 

Rather than use a float I was planning to allow a user to specify milliseconds via a tagged tuple:

```
{millisecond, 500}
```

...but that's different than what @rickard-green suggested [here](https://github.com/erlang/otp/issues/6283#issuecomment-1243578315).

**Comment by jdamanalo** [comment]
@lukebakken indeed, it is uncommon to have fractional minutes. I was just conscious of not adding a new "api". Nevertheless, I would be happy to implement the millisecond way. What do you think @rickard-green?

File: lib/os_mon/doc/src/disksup.xml [hunk 2]
```diff
@@ -63,7 +63,7 @@
     <p>The following configuration parameters can be used to change
       the default values for time interval and threshold:</p>
     <taglist>
-      <tag><c>disk_space_check_interval = int()>0</c></tag>
+      <tag><c>disk_space_check_interval = float()>=1/60000</c></tag>
```
Comment (rickard-green) [inline lib/os_mon/doc/src/disksup.xml:66]:
```suggestion
      <tag><c>disk_space_check_interval = number()>=1/60000</c></tag>
```

File: lib/os_mon/doc/src/disksup.xml [hunk 3]
```diff
@@ -126,7 +126,7 @@
       <name since="">set_check_interval(Minutes) -> ok</name>
       <fsummary>Set time interval, in minutes, for the periodic disk space check</fsummary>
       <type>
-        <v>Minutes = int()>=1</v>
+        <v>Minutes = float()>=1/60000</v>
```
Comment (rickard-green) [inline lib/os_mon/doc/src/disksup.xml:129]:
```suggestion
        <v>Minutes = number()>=1/60000</v>
```

**Review by rickard-green — CHANGES_REQUESTED** [review]
Regarding usage a `{millisecond, Time}` tuple. If so, I would rather want it to be `{TimeUnit, Time}`, where `TimeUnit` is of the type [`erlang:time_unit()`](https://www.erlang.org/doc/man/erlang.html#type-time_unit), than just the `millisecond` time unit. Also easy to implement using [`erlang:convert_time_unit/3`](https://www.erlang.org/doc/man/erlang.html#convert_time_unit-3) for conversion.

**Comment by lukebakken** [comment]
Thanks for the feedback @rickard-green

File: lib/os_mon/src/disksup.erl [hunk 4]
```diff
@@ -69,14 +69,23 @@ set_almost_full_threshold(Float) ->
...
     ok;
 dummy_reply(get_almost_full_threshold) ->
     round(os_mon:get_env(disksup, disk_almost_full_threshold) * 100);
 dummy_reply({set_almost_full_threshold, _}) ->
     ok.
 
+param_type(disk_space_check_interval, {millisecond, Time}) ->
+    param_type(disk_space_check_interval, Time);
+param_type(disk_space_check_interval, {second, Time}) ->
+    param_type(disk_space_check_interval, Time);
```
Comment (rickard-green) [inline lib/os_mon/src/disksup.erl:88]:
```suggestion
param_type(disk_space_check_interval, {TimeUnit, Time}) ->
    try erlang:convert_time_unit(Time, TimeUnit, millisecond) of
        MsTime when MsTime > 0 -> true;
        _ -> false
    catch
        _:_ -> false
    end;
```
Support for all time units. Note that code has not been tested.

File: lib/os_mon/doc/src/disksup.xml [hunk 5]
```diff
@@ -57,15 +57,40 @@
...
+            <p>The time interval in minutes.</p>
+          </item>
+          <tag><c>{second, int() >= 1}</c></tag>
+          <item>
+            <p>The time interval in seconds.</p>
+          </item>
+          <tag><c>{millisecond, int() >= 1}</c></tag>
+          <item>
+            <p>The time interval in milliseconds.</p>
+          </item>
```
Comment (rickard-green) [inline lib/os_mon/doc/src/disksup.xml:77]:
```suggestion
          <tag><c>{TimeUnit, Time}</c></tag>
          <item>
            <p>
              The time interval <c>Time</c> in a time unit specified by <c>TimeUnit</c> where
              <c>TimeUnit</c> is of the type
              <seetype marker="erts:erlang#time_unit/0"><c>erlang:time_unit/0</c></seetype>
              and <c>Time</c> is a positive integer. The time interval needs to be at least one
              millisecond long.
            </p>
          </item>
```
I see no reason to limit time unit to `second` and `millisecond`. I think it is much better to use the already defined time unit type.

Note that I've not tested building the docs using the above so the doc build might complain about it.

File: lib/os_mon/doc/src/disksup.xml [hunk 6]
```diff
@@ -57,15 +57,40 @@
       longer valid.</p>
   </description>
 
+  <datatypes>
+    <datatype>
+      <name>time()</name>
+      <desc>
+        <p>Supported units:</p>
+        <taglist>
+          <tag><c>int()</c></tag>
```
Comment (rickard-green) [inline lib/os_mon/doc/src/disksup.xml:66]:
```suggestion
          <tag><c>integer() >= 1</c></tag>
```

**Review by rickard-green — CHANGES_REQUESTED** [review]

File: lib/os_mon/doc/src/disksup.xml (hunk 5 above)
Comment (jdamanalo) [reply lib/os_mon/doc/src/disksup.xml]: This is better. I was initially worried of silently `floor()`ing values with parts of a millisecond but I see now that it just needed better documentation.

File: lib/os_mon/doc/src/disksup.xml [hunk 7]
```diff
@@ -57,15 +57,42 @@
...
+
   <section>
     <marker id="config"></marker>
     <title>Configuration</title>
     <p>The following configuration parameters can be used to change
       the default values for time interval and threshold:</p>
     <taglist>
-      <tag><c>disk_space_check_interval = int()>0</c></tag>
+      <tag>
+        <c>disk_space_check_interval = <seetype marker="#time">time()</seetype></c>
```
Comment (rickard-green) [inline lib/os_mon/doc/src/disksup.xml:92]:
```suggestion
        <c>disk_space_check_interval = </c><seetype marker="#time"><c>time()</c></seetype>
```
Doc build fails on this:
```
 ../xml/disksup.xml:92: element c: validity error : Element seetype is not declared in c list of possible children
      <c>disk_space_check_interval = <seetype marker="#time">time()</seetype></c
                                                                               ^
make[3]: *** [/buildroot/otp/make/otp_release_targets.mk:177: xmllint] Error 4
```
The above suggestion should solve it

**Review by rickard-green — CHANGES_REQUESTED** [review]

File: lib/os_mon/doc/src/disksup.xml (hunk 7 above)
Comment (rickard-green) [reply lib/os_mon/doc/src/disksup.xml]: I committed the fix, so we'll get this into the tests

**Review by rickard-green — APPROVED** [review]
I squashed and force pushed the doc fix.

We'll have a technical board meeting on Thursday where this PR will be brought up. I expect it to be accepted assuming we have no issues in testing.

**Pull request merged**